libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -1191,12 +1191,34 @@ prettyLinear = vsep' . map pretty
 prepareLinearization ::
   forall fn. BaseUniverse fn => Pred fn -> GE (SolverPlan fn)
 prepareLinearization p = do
-  let preds = flattenPred p
+  let preds = concatMap saturatePred $ flattenPred p
       hints = computeHints preds
       graph = transitiveClosure $ hints <> respecting hints (foldMap computeDependencies preds)
   plan <- linearize preds graph
   pure $ backPropagation $ SolverPlan plan graph
 
+-- TODO: generalize this to make it more flexible and extensible
+--
+-- The idea here is that we turn constraints into _extra_ constraints. C.f. the
+-- `mapIsJust` example in `Constrained.Examples.Map`:
+
+--    mapIsJust :: Specification BaseFn (Int, Int)
+--    mapIsJust = constrained' $ \ [var| x |] [var| y |] ->
+--      assert $ cJust_ x ==. lookup_ y (lit $ Map.fromList [(z, z) | z <- [100 .. 102]])
+
+-- Without this code the example wouldn't work because `y` is completely unconstrained during
+-- generation. With this code we essentially rewrite occurences of `cJust_ A == B` to
+-- `[cJust A == B, case B of Nothing -> False; Just _ -> True]` to add extra information
+-- about the variables in `B`. Consequently, `y` in the example above is
+-- constrained to `MemberSpec [100 .. 102]` in the plan.
+saturatePred :: forall fn. Pred fn -> [Pred fn]
+saturatePred p =
+  p
+    : case p of
+      Assert _ (Eql (FromG (SLeft _)) t) -> [toPreds t (SumSpec Nothing TrueSpec (MemberSpec []))]
+      Assert _ (Eql (FromG (SRight _)) t) -> [toPreds t (SumSpec Nothing (MemberSpec []) TrueSpec)]
```

Comment:
Could the MemberSpec[]  be FalseSpec?  I find MemberSpec[] to be problematic, since it can arise in so many ways.